### PR TITLE
[INJIMOB-3414]: Unsupported public key type

### PR DESCRIPTION
**Pull request by balachandarg-tw**
- Supporting only publicKeyMultibase

### CODE REVIEW

**Review by swatigoel — CHANGES_REQUESTED** [review]

File: Package.swift [hunk 1]
```diff
@@ -17,16 +17,20 @@ let package = Package(
         .package(url: "https://github.com/beatt83/jose-swift.git", .upToNextMinor(from: "4.0.2")),
         .package(url: "https://github.com/Alamofire/Alamofire.git", .upToNextMajor(from: "5.10.0")),
         .package(url: "https://github.com/valpackett/SwiftCBOR.git",  .upToNextMajor(from: "0.5.0")),
+        .package(url: "https://github.com/keefertaylor/Base58Swift.git", from: "2.1.0")
```
[the remark this answers is not in the document]
Comment (balachandarg-tw) [reply Package.swift]: Updated to range to achieve this.

File: Package.swift [hunk 2]
```diff
@@ -14,19 +14,23 @@ let package = Package(
             targets: ["OpenID4VP"]),
     ],
     dependencies: [
-        .package(url: "https://github.com/beatt83/jose-swift.git", .upToNextMinor(from: "4.0.2")),
-        .package(url: "https://github.com/Alamofire/Alamofire.git", .upToNextMajor(from: "5.10.0")),
-        .package(url: "https://github.com/valpackett/SwiftCBOR.git",  .upToNextMajor(from: "0.5.0")),
+        .package(url: "https://github.com/beatt83/jose-swift.git", "4.0.2"..<"4.0.3"),
+        .package(url: "https://github.com/Alamofire/Alamofire.git", "5.10.0"..<"5.10.1"),
```
Comment (KiruthikaJeyashankar) [inline Package.swift:18]: shall we keep Alamofire version as 5.10.2 same as previous one?

File: Sources/OpenID4VP/OpenID4VP.swift [hunk 3]
```diff
@@ -128,6 +128,11 @@ public class OpenID4VP {
         }
 
         errorInfo.merge(resolvedError.toErrorResponse()) { _, new in new }
+
```
Comment (KiruthikaJeyashankar) [inline Sources/OpenID4VP/OpenID4VP.swift:131]: Shall we also add some testcase around this?
- if state is available in auth request send back in error response as well?

File: Sources/OpenID4VP/Exceptions/OpenId4VPExceptions.swift [hunk 4]
```diff
@@ -146,6 +146,13 @@ class PublicKeyExtractionFailed: OpenID4VPException {
     }
 }
 
+class UnsupportedPublicKeyType: OpenID4VPException {
+    init(className: String) {
+        let message = "Unsupported Public Key type. Must be 'publicKeyMultibase'"
```
Comment (KiruthikaJeyashankar) [inline Sources/OpenID4VP/Exceptions/OpenId4VPExceptions.swift:151]: 🤔 Can we update the error message like only publicKeyMultibase is supported rather than Must be 'publicKeyMultibase`? Asking since the word must have been used for other spec related error messages as well (eg - "request must not be signed for clientid scheme did")
Like
```
Unsupported Public Key type. Supported: publicKeyMultibase
```

File: Sources/OpenID4VP/jwt/KeyResolver/types/DidPublicKeyResolver.swift [hunk 5]
```diff
@@ -30,19 +32,31 @@ class DidPublicKeyResolver : PublicKeyResolver {
...
+            for method in verificationMethods {
+                if let id = method["id"] as? String, id == kid {
+
+                    if !Self.supportedPublicKeyTypes.contains(where: { method[$0] != nil }) {
+                        throw UnsupportedPublicKeyType(className: DidPublicKeyResolver.className)
+                    }
+
+                    let publicKeyMultibase = method["publicKeyMultibase"] as? String
+                    if publicKeyMultibase?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true {
+                        throw InvalidData(
```
Comment (KiruthikaJeyashankar) [inline Sources/OpenID4VP/jwt/KeyResolver/types/DidPublicKeyResolver.swift:45]: Can we also add some tests for this - publicKeyMultibase cannot be null or empty

File: Tests/OpenID4VPTests/jwt/KeyResolver/DidPublicKeyResolverTests.swift [hunk 6]
```diff
@@ -64,13 +87,49 @@ class DidPublicKeyResolverTests : XCTestCase {
...
+                  "\(key)": "dummy-value"
+                }
+              ]
+            }
+            """
+            mockNetworkManager.setMockResponse(for: didDocumentUrl, responseBody: didDocumentWithUnsupportedKey)
+
+            let resolver = DidPublicKeyResolver(didUrl: did, networkManager: mockNetworkManager)
+
+            do {
```
Comment (KiruthikaJeyashankar) [inline Tests/OpenID4VPTests/jwt/KeyResolver/DidPublicKeyResolverTests.swift:122]: Can we use our custom assertAsyncThrowsError here?

**Review by KiruthikaJeyashankar — CHANGES_REQUESTED** [review]

File: Package.swift (hunk 2 above)
Comment (balachandarg-tw) [reply Package.swift:18]: Previously it was `5.10.0` only right ?

File: Sources/OpenID4VP/OpenID4VP.swift (hunk 3 above)
Comment (balachandarg-tw) [reply Sources/OpenID4VP/OpenID4VP.swift:138]: Done

**Review by KiruthikaJeyashankar — APPROVED** [review]

File: Tests/OpenID4VPTests/jwt/KeyResolver/DidPublicKeyResolverTests.swift [hunk 7]
```diff
@@ -1,76 +1,191 @@
...
+        "did:web:mosip.github.io:inji-mock-services:openid4vp-service:docs#key-0"
+      ],
+      "service": [],
+      "verificationMethod": [
+        {
+          "@context": "https://w3id.org/security/suites/ed25519-2020/v1",
+          "id": "did:web:mosip.github.io:inji-mock-services:openid4vp-service:docs#key-0",
+          "type": "Ed25519VerificationKey2020",
+          "controller": "did:web:mosip.github.io:inji-mock-services:openid4vp-service:docs",
+          "publicKeyMultibase": "z3CSkXmF1DmgVuqPFKMTuJgn846mEuVB9rNoyP9hXribo"
```
Comment (swatigoel) [inline Tests/OpenID4VPTests/jwt/KeyResolver/DidPublicKeyResolverTests.swift:28]: This test will fail as we have changed did document to have `publicKey`
Instead of resolving actual did document, let's mock the did document as done here https://github.com/mosip/inji-openid4vp/pull/106/files#diff-12effe492290188207ca292a4e9962bbda0aaea2aa11ee226de42dabc2d738ba

**Review by swatigoel — CHANGES_REQUESTED** [review]

**Review by swatigoel — APPROVED** [review]
